<a href="https://colab.research.google.com/github/navroz-lamba/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_t_tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [1]:

### YOUR WORK HERE
!curl https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data


3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,168.80,64.10,48.80,2548,dohc,four,130,mpfi,3.47,2.68,9.00,111,5000,21,27,13495
3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.60,168.80,64.10,48.80,2548,dohc,four,130,mpfi,3.47,2.68,9.00,111,5000,21,27,16500
1,?,alfa-romero,gas,std,two,hatchback,rwd,front,94.50,171.20,65.50,52.40,2823,ohcv,six,152,mpfi,2.68,3.47,9.00,154,5000,19,26,16500
2,164,audi,gas,std,four,sedan,fwd,front,99.80,176.60,66.20,54.30,2337,ohc,four,109,mpfi,3.19,3.40,10.00,102,5500,24,30,13950
2,164,audi,gas,std,four,sedan,4wd,front,99.40,176.60,66.40,54.30,2824,ohc,five,136,mpfi,3.19,3.40,8.00,115,5500,18,22,17450
2,?,audi,gas,std,two,sedan,fwd,front,99.80,177.30,66.30,53.10,2507,ohc,five,136,mpfi,3.19,3.40,8.50,110,5500,19,25,15250
1,158,audi,gas,std,four,sedan,fwd,front,105.80,192.70,71.40,55.70,2844,ohc,five,136,mpfi,3.19,3.40,8.50,110,5500,19,25,17710
1,?,audi,gas,std,four,wagon,fwd,front,105.80,192.70,71.40,55.70,2954,ohc,five,136,mpfi,3.19,3.40,8.50,1

In [2]:
import pandas as pd
from scipy import stats 
import numpy as np

In [3]:
column_names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 
                 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 
                 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
cars = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=column_names, na_values='?')
cars.head()

symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [4]:
cars_sample = cars.sample(10, random_state=30)
cars_sample.head()

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0

[5 rows x 26 columns]

In [5]:
print(cars['highway-mpg'].mean())
print(cars_sample['highway-mpg'].mean())

30.75121951219512
32.8


Sample being really small, the mean estimate of the sample and the entire dataset are quite different. 

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

1) Null Hypothesis :
All the cars have a fuel efficiency of about 35 miles per gallon on the highway.

$H_0: \overline{x} == \mu$

2) Aletrnative Hypothesis: 
$H_a: \overline{x} \neq \mu$ 

Cars in the lot have different fuel efficiency. 

3) Confidence Level: 95%

In [7]:
# with 10 samples
stats.ttest_1samp(cars_sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

4) **Conclusion**: 
Based on a t-statistic of .889, and a p-value of .363, we fail to reject the null hypothesis that all the cars have a fuel efficiency of about 35 miles per gallon on the highway. 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [8]:
# with 100 samples 
cars_sample1 = cars.sample(100, random_state=30)
cars_sample1.head()

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0

[5 rows x 26 columns]

In [9]:
stats.ttest_1samp(cars_sample1['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

**Conclusion:**
Based on a t-statistic of -6.518, and a p-value ~0, we reject the null hypothesis that all the cars have a fuel efficiency of about 35 miles per gallon on the highway, and suggest the alternative that there are differing levels of fuel efficiency in cars on the lot. 

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

It lead to different conclusions about the salesman's claim because of the the difference in sample size that we considered.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [10]:

### YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-07-14 03:50:58--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-07-14 03:50:59 (128 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [11]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers, na_values="?")

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [12]:
df = df.replace({'y': 1, 'n': 0})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [13]:
rep = df[df['party']=='republican']
dem = df[df['party']=='democrat']

In [14]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [16]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

1) Null Hypothesis: 
Democrats support this bill at the same level that republicans do.

$H_0: \overline{x}_r == \overline{x}_d$

2) Alternative Hypothesis: Democrats support this bill in a different proportion than republicans.

$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 99%

In [17]:
stats.ttest_ind(dem['budget'], rep['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

4) **Conclusion**: 
Due to a t-statistic of 23.2 and a p-value of ~0 we **reject** the null hypothesis that democrats and repulicans support the budget bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties.

Due to the t-statistic value being +ve democrats support this bill at a higher rate. 

In [ ]:
# Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level)

1) Null Hypothesis: 
Democrats support this bill at the same level that republicans do.

$H_0: \overline{x}_r == \overline{x}_d$

2) Alternative Hypothesis: Democrats support this bill in a different proportion than republicans.

$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 99%

In [18]:
stats.ttest_ind(dem['education'], rep['education'], nan_policy='omit')

Ttest_indResult(statistic=-20.500685724563073, pvalue=1.8834203990450192e-64)

4) **Conclusion**: 
Due to a t-statistic of -20.5 and a p-value of ~0 we **reject** the null hypothesis that democrats and repulicans support the education bill at the same rate, and suggest the alternative that there are differing levels of support for this bill between the two parties.

Due to the t-statistic value being -ve republicans support this bill at a higher rate. 

In [19]:
# Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

1) Null Hypothesis: 
Democrats support this bill at the same level that republicans do.

$H_0: \overline{x}_r == \overline{x}_d$

2) Alternative Hypothesis: Democrats support this bill in a different proportion than republicans.

$H_a: \overline{x}_r \neq \overline{x}_d$

3) Confidence Level: 90%

In [20]:
stats.ttest_ind(dem['immigration'],rep['immigration'], nan_policy='omit')

Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

4) **Conclusion**: 
Due to a t-statistic of -1.73 and a p-value of 0.083 we **fail to reject** the null hypothesis that democrats and repulicans support the education bill at the same rate. 

Due to the t-statistic value being -ve republicans support this bill at a slightly higher rate than democrats. 

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [21]:
# drop the attribute 'party'
dem1 = dem.drop('party', axis=1)
rep1 = rep.drop('party', axis=1)

In [22]:
# defining a function
def ttest(df1,df2):

  for col in df1:
    ttest = stats.ttest_ind(df1[col], df2[col], nan_policy='omit')
  
    print(f'{col}: {ttest}\n')

In [23]:
# List Comprehension 
ttest_list = [ttest(dem1,rep1) for col in (dem1,rep1)]

handicapped-infants: Ttest_indResult(statistic=9.205264294809222, pvalue=1.613440327937243e-18)

water-project: Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

budget: Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

physician-fee-freeze: Ttest_indResult(statistic=-49.36708157301406, pvalue=1.994262314074344e-177)

el-salvador-aid: Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

religious-groups: Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

anti-satellite-ban: Ttest_indResult(statistic=12.526187929077842, pvalue=8.521033017443867e-31)

aid-to-contras: Ttest_indResult(statistic=18.052093200819733, pvalue=2.82471841372357e-54)

mx-missile: Ttest_indResult(statistic=16.437503268542994, pvalue=5.03079265310811e-47)

immigration: Ttest_indResult(statistic=-1.7359117329695164, pvalue=0.08330248490425066)

synfuels: Ttest_indResult(statistic=8.293603989407588, pvalue=1.5759322



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

1) Null Hypothesis :
All the cars have a fuel efficiency of about 35 miles per gallon on the highway.

$H_0: \overline{x} == \mu$

2) Aletrnative Hypothesis: 
$H_a: \overline{x} \neq \mu$ 

Cars in the lot have different fuel efficiency. 

3) Confidence Level: 95%

In [24]:
# sample size = 50
n = 10

In [25]:
concept_of_population = 35 

In [26]:
# considering sample of 50 cars 
cars_sample2 = cars.sample(n, random_state=30) 

In [27]:
# calculating sample mean 
x_bar = cars_sample2['highway-mpg'].mean()
x_bar

32.8

In [29]:
# calculating Ho
h_o = x_bar - concept_of_population
h_o

-2.200000000000003

In [30]:
import numpy as np

# calculating T
t = h_o / (np.std(cars_sample2['highway-mpg']) / np.sqrt(n))
t

-1.0087922981253084

P value is 0.3253 

Since p > 0.5 we fail to reject the Null Hypothesis.

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE